scripts/generate-hotfix-metadata/generate-hotfix-metadata.py

```diff
@@ -0,0 +1,332 @@
+    #               \          /
+    #                DLPX-48962 -- HF-352
+    #
+    #
+
+    # Test that we correctly get the linked DLPX from a hotfix
+
+    results = get_linked_dlpx_to_hotfix(jira.issue('HF-351'))
+    result_keys = [r.key for r in results]
+    assert set(result_keys) == set(['DLPX-46713', 'DLPX-48962']), 'Error getting linked DLPX for HF-351'
+
+    results = get_backports(jira.issue('DLPX-49264'))
+    result_keys = [r.key for r in results]
+    assert set(result_keys) == set(['DLPX-49266', 'DLPX-72007']), \
+      'Error getting linked backport for DLPX-49264'
+  except Exception as e:
+
+    print ('Unexpected error:', e)
+    return False
+  return True
+
+
+def main():
+  parser = argparse.ArgumentParser()
+  parser.add_argument('-t', '--test',
+                      help='Run only the sanity tests (ignore all other options)'
+                      , action='store_true')
+  parser.add_argument('-v', '--version',
+                      help='The version you want to add in the form of x.x.x.x'
+                      )
+  parser.add_argument('-o', '--outfile',
+                      help='The file to save the generated hotfix metadata for the given version'
+                      , default='/tmp/hf_new')
+
+  args = parser.parse_args()
+  test = args.test
+  version = args.version
+  outfile = args.outfile
+
+  # Create an empty file if we do not have a version to avoid having to check for file existence in
+  # scripts that consume the file.
+  if version == "trunk" or not version:
+    with open(outfile, 'a') as f:
+      f.write("")
+    exit()
+
+  # Parse command line args
+
+  if args.test:
+    if not run_sanity_tests():
+      die('Sanity tests failed')
+    print('Sanity tests passed')
+    sys.exit(0)
+  else:
+    if version is None or outfile is None:
+      parser.error('if -t is not given, -v and -o are required')
+
+
+  # We allow some flexibility in version strings. So long as the version
+  # string starts with what looks like a version we accept it
+  # so that we can accept version strings that represent an
+  # unreleased appliance version (e.g.
+  # 6.1.0.0-snapshot.20210706184036139...)
+  match=re.search("^[0-9]+\.[0-9]+\.[0-9]+.[0-9]+", version)
+
+  if match:
+    version = match.group()
+  else:
+    die("version must start with 4 numerals separated by '.'")
+
+  # Check if appropriate environment variables are set so that we can query Jira
+
+  JIRA_URL = os.getenv('JIRA_URL', None)
+  JIRA_USER = os.getenv('JIRA_USER', None)
+  JIRA_PASSWORD = os.getenv('JIRA_PASSWORD', None)
+
+  # Gradle will pass in the string "null" if the values are not set
+  errors = []
+  if not JIRA_URL or JIRA_URL == 'null':
+    errors.append("Environment variable JIRA_URL not set.")
+  if not JIRA_USER or JIRA_USER == 'null':
+    errors.append("Environment variable JIRA_USER not set.")
+  if not JIRA_PASSWORD or JIRA_PASSWORD == 'null':
+    errors.append("Environment variable JIRA_PASSWORD not set.")
```

Comment:
Is appliance-build going to fail when run manually if those environment variables are not set?

We are trying to make it easier for users to run appliance-build manually, so they can do so without having to set a bunch of environment variables; the build skips the steps that requires those variables to be set. At the same time we want to make sure that all of those variables are passed when run by automation to avoid silent regressions (see for instance https://github.com/delphix/appliance-build/pull/570). Perhaps you could add a similar check to the calling `run.sh` script (it could generate an empty hotfix file so that generation of delphix-entire doesn't fail)?

Reply:
A build without hotfixmetadata is not a valid build as I would assume other cases where steps are skipped due to missing inputs. Do we provide a means to tell the user what they are missing by not providing these addtional environment variables? For example, if we do not provide the environment variables for querying Jira to get hotfix metadata, then the resulting image should not be used for anything related to hotfixes.

What about providing flags to bypass features that require these environment variables. When these bypasses are not provided and the environment variables are not present we must then require the error message explain the flags needed to bypass reading the environment variable. This approach makes it is clear to users of appliance-build that they are intentionally running appliance-build in such a way that it may generate an image that may not work as intended, such as having missing features in the product. 

I am concerned that if the default behavior is to build an image that has missing components without knowing what is missing, there are going to be issues with using the image because incorrect assumptions are going to be made. I also don't think reading through the build log is sufficient information to convey to the user what is skipped unless we also have an annoying summary message at the end saying what was skipped.

Reply:
Yeah we have the same issue with a bunch of other environment variables, for instance the signature, the delphix appliance version, etc.

I think it would be reasonable to check for all the environment variables in the beginning of the build and issue clear warnings if they are not set. However I believe it would be a regression to fail the build if some of those variables are not set when running the build manually. Note that if we use similar checks to https://github.com/delphix/appliance-build/pull/570 we would not be able to accidentally produce a broken production image.

Reply:
FWIW, if appliance-build is currently run outside of Jenkins, and the user does not properly construct their environment to contain the all of the "special" envirionment variables, then the image will not work when used with the product's upgrade UI; this is already the case today.

So while I understand the sentiment, that we should try to produce an image that will work properly, that goal conflicts with the other goal of being able to run the build outside of Jenkins, and not requiring the user to need intimate knowledge about how to set up the environment properly (i.e. the build should work with the default environment).

IMO, if the JIRA environment variables are not present, and we're not building via Jenkins, than I think it's reasonable to produce an image that might not be entirely "correct", such that it doesn't have the proper hotfix metadata. When run outside of Jenkins, the priority should be to allow the build to succeed, while performing as much of the normal build logic as we can, without hindering the user-interface of the build for that manual invocation (i.e. without forcing the user to specify lots of environment variables properly). Further, it's less of a priority to ensure the manually generated artifacts match the Jenkins produced artifacts.

I think of a build run outside of Jenkins more of a best-effort type of thing, rather than necessarily having to product artifacts that will work properly when used with the product.

> there are going to be issues with using the image because incorrect assumptions are going to be made.

The artifacts that are generated outside of Jenkins, shouldn't really be used for anything beyond manual testing and/or developer iteration, and the developer that built the artifacts will need to know what is an appropriate test with the manually generated artifacts, and what is not appropriate. So, I think the concern is valid, but we've already adopted the notion that it's up to the developer to ensure they don't use the artifacts incorrectly, so I don't think we should be changing that approach in the context of this PR.

For example, if we did want to change that approach, we should do it holistically, and look at other areas of the build that will not (currently) work "correctly" when missing environment variables, which I (also) don't think is appropriate for this PR.

Reply:
I am suggesting is to either force the user to supply the correct parameters to skip parts of the build  (e.g. "without flags") or to provide a summary of what was not included in the build and what environment variables would be required to enable the feature, similar to how many open source software reports what features are missing at the end of a compile or configure (E.g. `./configure` or `make`). 

While we do not want to require users to have detailed knowledge of which environment variables to set to run a build, we might want to make it a little easier for a developers to become more knowledgable without having to hunt through the code or the devops code just to run a build a certain way. 

Of course all of this is out of scope for this review.

Reply:
Yea, I'm not against making things more clear as you suggest. I'd expect that to be a decent amount of effort, so doing this in another PR makes sense (if it's going to be done, at all).

Reply:
> I am suggesting is to either force the user to supply the correct parameters to skip parts of the build (e.g. "without flags") or to provide a summary of what was not included in the build and what environment variables would be required to enable the feature, similar to how many open source software reports what features are missing at the end of a compile or configure

That's reasonable and would probably be helpful. I agree with Prakash that it would require a bit of extra work though, so I'm OK deferring this to another PR.